Adds setting to toggle the WCPay Subscriptions feature

mattallan opened the pull request:
Fixes #3064 

### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

- **New setting to toggle WCPay Subscriptions**

This PR adds a new setting to allow Store managers to enable/disable WCPay Subscriptions.

![](https://d.pr/i/QOJoFl+)

Once https://github.com/Automattic/woocommerce-payments/issues/3087 is implemented, WCPay Subscriptions will be enabled by default for stores based in the US, so we want to have an option for store managers to disable this if they don't want it.

With this setting disabled, `woocommerce-subscriptions-core` will not be loaded and all built-in Subscriptions functionality will be turned off.

At the moment this toggle setting is being displayed when either is true:
 - Stores base location is set to the US
 - Store already has the WCPay Subscriptions feature enabled

### Questions

1. Should we show this setting if the store already has WC Subscriptions (the extension)?
2. Is the wording okay?
2. What documentation link should we send store managers to? At the moment I've got it set to a link that doesn't exist yet (https://docs.woocommerce.com/document/payments/subscriptions/)
3. Is the order of these advanced settings okay?

I'll chase these questions up internally (pdjTHR-er-p2)

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

Checkout this branch and make sure assets are built before testing

1. Delete the `_wcpay_feature_subscriptions` feature flag if you already have it set
2. Confirm that Subscriptions isn't enabled (no **WooCommerce > Subscriptions**)
2. Make sure your store's base country is set to US (**WooCommerce > Settings > General**)
3. Visit WooCommerce > Settings > Payments
4. Open the advanced settings
5. Tick the box to **Enable Subscriptions with WooCommerce Payments** and save the settings
6. Confirm Subscriptions is now enabled on the store
7. Uncheck the **Enable Subscriptions with WooCommerce Payments** and save the settings
8. Confirm Subscriptions is disabled

---

Closes #3064

haszari commented:
> At the moment this toggle setting is being displayed when either is true:
> - Stores base location is set to the US
> - Store already has the WCPay Subscriptions feature enabled

@mattallan I'm thinking we should show the checkbox on all stores to keep the logic simple and the UI consistent, unless there's specific reasons to hide it. cc @james-allan who may be more up to date on discussions around this than me? 😄 

i.e. the checkbox to preview the feature is available to all, and it is checked by default for US-based stores.

Related – should we adjust the wording to mention that this feature is "preview" stage? As well as linking to the doc to provide full details.

haszari commented:
Just saw the questions @mattallan – good questions! Here's my take, keen to hear from @james-allan 👍 and maybe get a design opinion. Let's get this in the best shape we can (wording etc) and then seek feedback from design/product.

> Should we show this setting if the store already has WC Subscriptions (the extension)?

Hmm, not sure of a precedent for this. I suppose the option is not available if subs is installed, so maybe disabled with an explanation?

> Is the wording okay?

Looks good though maybe should mention preview status - i.e. this is early release. Though we'll need to update when this goes wider/full support.

> What documentation link should we send store managers to? At the moment I've got it set to a link that doesn't exist yet (https://docs.woocommerce.com/document/payments/subscriptions/)

Based on the plans for docs, I think we have a couple of options.

1. High level `Subscriptions` doc that outlines that subs is available as a flexible extension and as a feature within WC Pay.
2. Explainer/differences doc that outlines the differences between the two offerings.

> Is the order of these advanced settings okay?

Looks good to me – this makes sense at the bottom of the non-dev settings since it's newest and preview status.

mattallan commented:
I've moved this PR from draft to ready for review. The only thing that came up from internal discussions is that this setting shouldn't be shown to stores that already have the subscriptions extension active. Please see pdjTHR-er-p2#comment-241.

I've updated this PR to not show this setting when Subscriptions is enabled.

---

@haszari thanks a bunch for providing an initial round of feedback on this one!
Based on Brents latest comments on the p2 (see link above) let me know if there's anything you think we should still change :)

> Looks good though maybe should mention preview status - i.e. this is early release. Though we'll need to update when this goes wider/full support.

To avoid having to change this wording down the track and then do another wcpay release, as well as updating any screenshots we use in documentation, I think it makes sense to not mention on the settings page that this is a preview feature.

We can use the external documentation link to explain this is only available to US at the moment.

james-allan reviewed (approved):
From a code and functionality perspective, this looks good to me. 

FWIW I tested the following additional scenarios: 

1. If you are not eligible (not located in the US) but have enabled WCPay Subscriptions via the database or via WCPay Dev tools plugin, disabling the checkbox makes it disappear immediately. Saving does still disable the feature. Not sure if this is the intended UX but given you'd have to jump through a number of hoops as a customer to run into that, it's fine. It might even be intended. If you previously enabled the feature via the database, you can disable it via the UI but it wont allow you to re-enable it via the UI once you uncheck it.
2. If I have the feature enabled (via dev tools, this checkbox, or the database), you have the Subscriptions plugin active. Attempting to deactivate the subscriptions plugin will display the popup warning added in #3061. If the feature is disabled (via dev tools, this checkbox, or the database), then the popup isn't shown. This works as intended.

haszari commented:
> If you are not eligible (not located in the US) but have enabled WCPay Subscriptions via the database or via WCPay Dev tools plugin, disabling the checkbox makes it disappear immediately. Saving does still disable the feature. Not sure if this is the intended UX but given you'd have to jump through a number of hoops as a customer to run into that, it's fine. It might even be intended. If you previously enabled the feature via the database, you can disable it via the UI but it wont allow you to re-enable it via the UI once you uncheck it.

@mattallan - Hiding the checkbox for non-US stores seems unnecessary to me - and more complex. Can we show for all stores to keep it simple? (also commented on P2)
